### PR TITLE
Fixes all known NameError/ValueError bugs when jitting

### PR DESCRIPTION
**Pull request by dwillmer**
Needs rebasing, but fixes all of the known issues causing NameError/ValueError, and adds unit-tests to cover all the permutations of nested functions.

### CODE REVIEW

**Comment by dwillmer** [comment]
Will fix up the codacy issues, but it's ready for review.

**Comment by dwillmer** [comment]
@rjenc29 @pawroman - please review!

I'm happy to ignore the codecov/patch failure given the codecov/project coverage has increased.

This solves the NameError issues we were seeing + others. There's various further updates we can make (such as removing the context manager from decorator.py), but i felt this was big enough already.

**Comment by pawroman** [comment]
Coverage issues can be addressed by my `pragma: no cover` thing - currently in #18

File: fastats/core/ast_transforms/convert_to_jit.py [hunk 1]
```diff
@@ -8,7 +8,8 @@ def convert_to_jit(func):
         return func
 
     _jit = jit(nopython=True, nogil=True)
-    return _jit(func)
+    rv = _jit(func)
+    return rv
```
Comment (pawroman) [inline fastats/core/ast_transforms/convert_to_jit.py]: Why the temp var? Debugging purposes?

File: fastats/core/ast_transforms/processor.py [hunk 2]
```diff
@@ -22,55 +23,81 @@ def __init__(self, top_level_func, overrides, replaced, new_funcs=None):
         self._debug = self._overrides.get('debug')
 
     def process(self):
-        source = inspect.getsource(self.top_level_func)
-        tree = ast.parse(source)
+        try:
+            source = inspect.getsource(self.top_level_func)
+            tree = ast.parse(source)
+        except IndentationError:
```
Comment (pawroman) [inline fastats/core/ast_transforms/processor.py]: Where can this come from? `parse`?

File: fastats/core/ast_transforms/processor.py [hunk 3]
```diff
@@ -22,55 +23,81 @@ def __init__(self, top_level_func, overrides, replaced, new_funcs=None):
...
+            data = uncompile(self.top_level_func.__code__)
+            tree = parse_snippet(*data)
+
         globs = self.top_level_func.__globals__
+        globs['jit'] = jit
```
Comment (pawroman) [inline fastats/core/ast_transforms/processor.py]: This deserves a comment

File: fastats/core/ast_transforms/processor.py [hunk 4]
```diff
@@ -22,55 +23,81 @@ def __init__(self, top_level_func, overrides, replaced, new_funcs=None):
...
+    if c.co_name == '<lambda>':
+        raise TypeError('lambda functions not supported')
+    if c.co_filename == '<string>':
+        raise ValueError('code without source file not supported')
+
+    filename = inspect.getfile(c)
+    try:
+        lines, firstlineno = inspect.getsourcelines(c)
+    except IOError:
+        raise Exception('source code not available')
```
Comment (pawroman) [inline fastats/core/ast_transforms/processor.py]: `except IOError as err:` and `raise RuntimeError('source code not available') from err` might be more suitable

File: fastats/core/ast_transforms/transformer.py [hunk 5]
```diff
@@ -54,30 +57,38 @@ def visit_Call(self, node):
...
-            proc = AstProcessor(
-                orig_inner_func, self._params,
-                self._replaced, self._new_funcs
-            )
-            new_inner_func = proc.process()
-            self._globals[node.func.id] = convert_to_jit(new_inner_func)
+
+            not_ufunc = not isinstance(orig_inner_func, np.ufunc)
+            not_builtin = not isbuiltin(orig_inner_func)
+            if not_ufunc and not_builtin:
```
Comment (pawroman) [inline fastats/core/ast_transforms/transformer.py:80]: I think we should refactor out these checks to some common util function. See changes in #18 : https://github.com/pawroman/fastats/blob/7169f578892f9a72c2c14baa9bfd1ce2b7f9b9ec/fastats/core/ast_transforms/convert_to_jit.py#L15-L19

File: fastats/core/decorator.py [hunk 6]
```diff
@@ -16,9 +16,8 @@ def code_transform(func, replaced):
     try:
         yield func
     finally:
-        for k, v in replaced.items():
-            func.__globals__[k] = v
-        replaced.clear()
+        # TODO : remove this context manager as it's now not needed?
+        pass
```
Comment (pawroman) [inline fastats/core/decorator.py]: Definitely not needed anymore

File: fastats/core/decorator.py [hunk 7]
```diff
@@ -29,29 +28,48 @@ def fs(func):
...
 
         with suppress(KeyError):
             del kwargs['return_callable']
 
+        # This deliberately mutates the kwargs.
+        # We don't want to have a fs-decorated function
+        # as a kwarg to another, so we undecorate it first.
+        for k, v in kwargs.items():
+            if hasattr(v, 'undecorated'):
+                kwargs[k] = v.undecorated
```
Comment (pawroman) [inline fastats/core/decorator.py:22]: Would be nice to have helper funcs for marking/unmarking this

File: tests/core/ast_transforms/test_basic_sanity_deep_hierarchy.py [hunk 8]
```diff
@@ -31,7 +31,8 @@ def double(x):
 
 def triple(x):
     # TODO : investigate double(x) + x failure.
-    return 3*x
+    # return 3 * x
```
Comment (pawroman) [inline tests/core/ast_transforms/test_basic_sanity_deep_hierarchy.py]: Dead code

File: tests/test_name_error_addition.py [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+NameErrors occasionally when nested ast-transformed
+code was jitted using numba.
+
+The same code without jitting ran fine - the addition
+of the jit decorator caused the issues.
+
+Do not change the hierarchy below - it's set up to
+test some very specific edge cases:
+
+-
```
Comment (pawroman) [inline tests/test_name_error_addition.py]: Docstring formatting

File: tests/test_name_error_nested.py [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+def test_nested_basic_sanity():
+    """
+    Some trivially nested functions started
+    throwing name errors.
+    Just using single_pass(value=half) works
+    fine, the value() function gets correctly
+    replace with the half() function, however
+    single_pass(value=quarter) doesn't work,
+    it throws a NameError.
```
Comment (pawroman) [inline tests/test_name_error_nested.py]: throws -> raises

**Review by pawroman — CHANGES_REQUESTED** [review]
Looks good spare for minor issues

File: fastats/core/ast_transforms/convert_to_jit.py (hunk 1 above)
Comment (dwillmer) [reply fastats/core/ast_transforms/convert_to_jit.py]: Yep, precisely.

File: fastats/core/ast_transforms/processor.py (hunk 2 above)
Comment (dwillmer) [reply fastats/core/ast_transforms/processor.py]: Yep, precisely. `parse` will raise an IndentationError if a nested function source is passed in

File: fastats/core/ast_transforms/processor.py (hunk 3 above)
Comment (dwillmer) [reply fastats/core/ast_transforms/processor.py]: Adding it to the globals dict means that an explicit import isn't required for dynamically generated code. Can add it as a comment, i just thought it was obvious.

File: fastats/core/ast_transforms/processor.py (hunk 4 above)
Comment (dwillmer) [reply fastats/core/ast_transforms/processor.py]: hmm, i agree `raise ... from` is better here, but i don't like RuntimeError. Will have a think about this one.

File: fastats/core/ast_transforms/transformer.py (hunk 5 above)
Comment (dwillmer) [reply fastats/core/ast_transforms/transformer.py:80]: Yep, agreed. I suggest we ignore it here and take the refactor from your PR.

File: fastats/core/decorator.py (hunk 6 above)
Comment (dwillmer) [reply fastats/core/decorator.py]: Agreed - again, let's take the updated version from your PR

File: tests/core/ast_transforms/test_basic_sanity_deep_hierarchy.py (hunk 8 above)
Comment (dwillmer) [reply tests/core/ast_transforms/test_basic_sanity_deep_hierarchy.py]: Nice one, thanks

File: fastats/core/decorator.py (hunk 7 above)
Comment (dwillmer) [reply fastats/core/decorator.py:22]: My feeling is that splitting this code into separate functions would lose clarity. This only applies to the decorator it's defined in, so I think this is fine, unless there's a valid use case for the helper functions.

File: fastats/core/ast_transforms/processor.py (hunk 2 above)
Comment (pawroman) [reply fastats/core/ast_transforms/processor.py]: Let's keep the `try/except` limited to `parse` then

File: fastats/core/ast_transforms/processor.py (hunk 3 above)
Comment (pawroman) [reply fastats/core/ast_transforms/processor.py]: Nothing's obvious when you're doing AST rewrites ;)

File: fastats/core/decorator.py (hunk 7 above)
Comment (pawroman) [reply fastats/core/decorator.py:22]: OK - let's refactor if there's more metadata we need to store

File: tests/test_name_error_nested.py [hunk 11]
```diff
@@ -0,0 +1,51 @@
+"""
+This is a test for a specific edge case:
+
+if the function passed as a kwarg has a nested
+function, it doesn't get processed properly.
```
Comment (pawroman) [inline tests/test_name_error_nested.py:5]: How about past tense in these docstrings? It's a problem that's been fixed!

Comment (pawroman) [reply tests/test_name_error_nested.py:5]: BTW, good work on all these tests, really nice to have them

File: tests/core/test_single_pass.py [hunk 12]
```diff
@@ -127,18 +130,9 @@ def mean(x):
...
-    #     return sum(x) / len(x)
-    #
-    # result_py = single_pass(data, value=mean_py)
-    #
-    # assert result_py[0][0] == approx(0.5)
-    # assert result_py[4][0] == approx(8.5)
-
+    # len(x) works fine...
     def mean_npy(x):
-        return nsum(x) / len(x)
```
Comment (pawroman) [inline tests/core/test_single_pass.py:145]: Why has this stopped working?

**Comment by dwillmer** [comment]
Thanks for the reviews, working on this now.

**Comment by dwillmer** [comment]
Ok, I believe all the code review comments have been addressed, let's get this in.

**Review by pawroman — APPROVED** [review]
